전남대 BE 서영서 - Step2 회원 로그인

yeongseo1225 opened the pull request:
step2 회원 로그인
회원은 이메일과 비밀번호를 입력하여 가입한다.
토큰을 받으려면 이메일과 비밀번호를 보내야 하며, 가입한 이메일과 비밀번호가 일치하면 토큰이 발급된다.
토큰을 생성하는 방법에는 여러 가지가 있다. 방법 중 하나를 선택한다.
(선택) 회원을 조회, 추가, 수정, 삭제할 수 있는 관리자 화면을 구현한다.

시크릿정보를 애플리케이션 설정으로 빼고 나니 의존성 주입이 필요해져서 유틸리티 클래스에서 다시 빈으로 등록해주었습니다. 
현재는 단순한 구조이지만 추후 유효성 검사나 필드 변경 가능성이 생길 경우, class가 가독성과 유지보수 측면에서 더 유리할 것 같다는 생각이 들었습니다. 그래서 지금 record로 변경해두면, 나중에 다시 class로 리팩토링할 때 오히려 불편함이 생길 수 있을 것 같은데, 멘토님께서는 이 부분에 대해 어떻게 생각하시는지 궁금합니다

wotjd243 committed:
Initial commit

wotjd243 committed:
feat: set up the project

yeongseo1225 committed:
전남대 BE_서영서 0단계 - 기본 코드 준비 (#84)

* chore : 2단계 과제 진행을 위한 기존 코드 이관

* docs(readme) : step0 기능 구현 목록 추가

yeongseo1225 committed:
전남대_BE_서영서 2단계 step1 (#173)

* docs(readme) : step1 기능 목록 추가

* feat(deps) : 입력값 검증을 위한 validation 의존성 추가

* refactor(api): 요청 DTO 적용하여 ProductController 구조 개선

* feat : 상품명 유효성 검사 기능 추가

* test : ProductNameValidator 테스트 코드 추가

* feat: 유효성 검사 실패 시 에러 메시지를 JSON 형식으로 반환하는 핸들러 추가

* refactor: ProductController에서 도메인 대신 요청 DTO 사용하도록 리팩토링

* test: 유효하지 않은 상품 등록 요청 시 에러 응답 JSON 테스트 추가

* refactor : ProductRequest에 setter 제거 및 final 필드 적용으로 외부 상태 변경 차단

* refactor : Product 도메인 객체를 불변 구조로 전환. id는 JDBC 대응을 위해 setter 유지

* feat: 조회 전용 DTO ProductResponse 추가하여 도메인 객체 직접 노출 방지

* refactor : AdminProductController를 DTO 분리 구조에 맞게 리팩토링

* refactor: 예외 응답 구조를 DTO 기반으로 리팩토링하여 응답 일관성 확보

* refactor: 정규식 Pattern을 static으로 캐싱하여 ProductNameValidator 성능 개선

* feat : ValidProductName에 allowKakao 옵션 추가하여 유연한 상품명 검증 지원

yeongseo1225 committed:
feat : Memeber 도메인 및 JDBC repository 구현

yeongseo1225 committed:
build : jwt 라이브러리 의존성 추가

yeongseo1225 committed:
feat(member) : role 필드 추가 및 저장/조회 로직 반영

yeongseo1225 committed:
feat : JWT 토큰 생성 유틸 클래스 추가

yeongseo1225 committed:
회원 가입 요청 dto 추가

yeongseo1225 committed:
build : spring-security 의존성 추가

yeongseo1225 committed:
feat : 회원 등록 로직 구현 (토큰 발급 제외)

yeongseo1225 committed:
feat(security) : BCrypt 설정 및 기본 시큐리티 필터 구성

yeongseo1225 committed:
feat : 회원가입 시 토큰 발급 기능 추가

yeongseo1225 committed:
feat : 로그인 요청 DTO 추가

yeongseo1225 committed:
feat: 로그인 로직 구현

yeongseo1225 committed:
feat : 로그인 API 앤드포인트 추가

yeongseo1225 committed:
feat(error): 인증/인가 관련 예외 처리 및 상태 코드 응답 구현

yeongseo1225 committed:
chore: 의존성 정리 및 카테고리별 정렬 (Spring, JWT, DB, Test)

yeongseo1225 committed:
feat(security): JWT 인증 필터 구현 및 SecurityConfig에 필터 등록

yeongseo1225 committed:
feat: 비밀번호 찾기 및 변경 요청 시 403 Forbidden 반환 처리 추가

yeongseo1225 committed:
docs(readme) : step2 기능 구현 목록 추가

yeongseo1225 committed:
docs(readme) : step2 항목 포맷 수정

leaf-upper commented on src/main/java/gift/config/SecurityConfig.java:
```diff
@@ -0,0 +1,40 @@
+                        .requestMatchers("/api/members/register", "/api/members/login").permitAll()
+                        .anyRequest().permitAll()
+                )
+                .addFilterBefore(new JwtFilter(jwtUtil), UsernamePasswordAuthenticationFilter.class)
```
SpringSecurity를 사용하셨군요 👍 👍

leaf-upper commented on src/main/java/gift/util/JwtUtil.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Date;
+
+@Component
+public class JwtUtil {
```
Util클래스는 Bean으로 등록하기보다는 UtiltiyClass로써 사용하는게 좋지않을까요?
- 상속불가
- public static method
등으로 지원해줘도 괜찮지 않을까 싶어요

leaf-upper commented on src/main/java/gift/domain/Member.java:
```diff
@@ -0,0 +1,41 @@
+    private Long id;
+    private String email;
+    private String password;
+    private String role;
```
Role은 Enum으로 관리하는건 어떠세요?

leaf-upper commented on src/main/java/gift/dto/MemberRegisterRequest.java:
```diff
@@ -0,0 +1,17 @@
+package gift.dto;
+
+public class MemberRegisterRequest {
```
Request는 record와 같은 불변객체로 유지해도 될것 같아요.

leaf-upper commented on src/main/java/gift/security/JwtFilter.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request,
```
Filter레벨에서 UnauthorizedException 예외가 발생하면 GlobalExceptionHandler가 처리할수있을까요?

leaf-upper commented on src/main/java/gift/service/MemberService.java:
```diff
@@ -0,0 +1,28 @@
+            throw new IllegalArgumentException("이미 존재하는 이메일입니다.");
+        }
+
+        String encodedPw = passwordEncoder.encode(request.getPassword());
```
PasswordEncoder에 대해서 간략하게 설명해주실수 있으실까요?
어떤 알고리즘을 사용하여 비밀번호를 암호화할까요?

yeongseo1225 replied:
PasswordEncoder는 Spring Security에서 제공하는 인터페이스로 비밀번호를 안전하게 암호화하고 비교하기 위한 기능을 제공합니다. 가장 일반적으로 사용되는 BCryptPasswordEncoder를 사용하였고, 단반향 해시라서 암호화된 비밀번호를 복호화할 수 없고 같은 비밀 번호라도 해시 값이 매번 다르게 생성되어 보안성이 높다는 특징이 있습니다.

leaf-upper commented on src/main/java/gift/util/JwtUtil.java:
```diff
@@ -0,0 +1,54 @@
+@Component
+public class JwtUtil {
+
+    private final String secret = "D23kjibddFAWerws2AFlL8WXmohJMCvigQggaEypa5E=";
```
시크릿정보는 애플리케이션 설정으로 빼는건 어떠세요?

leaf-upper reviewed (changes requested):
안녕하세요 영서님
스프링 시큐리티를 이용해서 회원가입 기능을 구현해주셨네요~
시큐리티 코드쪽에 대해서 본인이 생각하시는 부분을 조금더 코멘트로 작성해주시면
리뷰해드리기 편할것 같아요~
몇가지 코멘트를 남겨드렸는데 확인한번 부탁드릴게요~

yeongseo1225 committed:
refactor: JwtUtil 클래스를 static 유틸리티 클래스로 리팩터링

- @Component 제거하여 스프링 Bean 등록 방지
- 클래스에 final 및 private 생성자 추가로 인스턴스화 및 상속 방지
- 모든 메서드를 public static으로 변경하여 전역 접근 가능하도록 수정

yeongseo1225 committed:
refactor: JwtUtil에 설정값 주입을 위해 @Component로 재등록 및 secret 외부화

yeongseo1225 committed:
refactor: Member의 role 필드를 Enum 타입으로 변경

yeongseo1225 committed:
refactor: JwtFilter에서 예외 발생 시 직접 JSON 응답 반환하도록 개선
